File: backend/app/services/diff_fetcher.py
import httpx

class DiffFetcher:
    def __init__(self, github_token: str):
        self.github_token = github_token

    def fetch_files(self, owner: str, repo: str, pr_number: int):
        url = f'https://api.github.com/repos/{owner}/{repo}/pulls/{pr_number}/files?per_page=100'
        headers = {'Authorization': f'Bearer {self.github_token}', 'Accept': 'application/vnd.github+json'}
        files = []
        with httpx.Client(timeout=30.0) as client:
            while url:
                r = client.get(url, headers=headers)
                r.raise_for_status()
                files.extend(r.json())
                link = r.headers.get('link', '')
                next_url = None
                for part in link.split(','):
                    if 'rel="next"' in part:
                        next_url = part.split(';')[0].strip().strip('<>')
                url = next_url
        return files

File: backend/src/services/diff-fetcher.ts
import axios, { AxiosResponse } from 'axios';
import { GITHUB_DOTCOM_API } from '../integrations/github/host.js';

/**
 * Fetches PR files from GitHub API with pagination support.
 */
export class DiffFetcher {
  private githubToken: string;
  private apiBaseUrl: string;

  constructor(githubToken: string, apiBaseUrl: string = GITHUB_DOTCOM_API) {
    this.githubToken = githubToken;
    this.apiBaseUrl = apiBaseUrl;
  }

  async fetchFiles(owner: string, repo: string, prNumber: number): Promise<Record<string, unknown>[]> {
    let url: string | null = `${this.apiBaseUrl}/repos/${owner}/${repo}/pulls/${prNumber}/files?per_page=100`;
    const headers = {
      'Authorization': `Bearer ${this.githubToken}`,
      'Accept': 'application/vnd.github+json',
    };
    const files: Record<string, unknown>[] = [];

    while (url) {
      const response: AxiosResponse = await axios.get(url, { headers, timeout: 30000 });
      files.push(...(response.data as Record<string, unknown>[]));

      // Parse Link header for pagination
      const linkHeader: string = (response.headers['link'] as string) || '';
      let nextUrl: string | null = null;
      for (const part of linkHeader.split(',')) {
        if (part.includes('rel="next"')) {
          const match: RegExpMatchArray | null = part.match(/<([^>]+)>/);
          if (match) {
            nextUrl = match[1];
          }
        }
      }
      url = nextUrl;
    }

    return files;
  }
}
